syntax = "proto3";

message WriteRequest {
  repeated TimeSeries timeseries = 1;
  reserved 2;
  reserved 3;
}

message TimeSeries {
  repeated Label labels = 1;
  repeated Sample samples = 2;
}

message Label {
  string name  = 1;
  string value = 2;
}

message Sample {
  double value    = 1;
  int64 timestamp = 2;
}

message ReadRequest {
  repeated Query queries = 1;

  enum ResponseType {
    SAMPLES = 0;
    STREAMED_XOR_CHUNKS = 1;
  }

  repeated ResponseType accepted_response_types = 2;
}

message ReadResponse {
  repeated QueryResult results = 1;
}

message LabelMatcher {
  enum Type {
    EQ  = 0;
    NEQ = 1;
    RE  = 2;
    NRE = 3;
  }
  Type type    = 1;
  string name  = 2;
  string value = 3;
}

message ReadHints {
  int64 step_ms = 1; 
  string func = 2;   
  int64 start_ms = 3;
  int64 end_ms = 4;  
  repeated string grouping = 5;
  bool by = 6;
  int64 range_ms = 7;
}

message Query {
  int64 start_timestamp_ms = 1;
  int64 end_timestamp_ms = 2;
  repeated LabelMatcher matchers = 3;
  ReadHints hints = 4;
}

message QueryResult {
  repeated TimeSeries timeseries = 1;
}
